pkg/core/gas_price.go

```diff
@@ -0,0 +1,106 @@
+package core
+
+import (
+	"github.com/CityOfZion/neo-go/pkg/smartcontract"
+	"github.com/CityOfZion/neo-go/pkg/util"
+	"github.com/CityOfZion/neo-go/pkg/vm"
+	"github.com/CityOfZion/neo-go/pkg/vm/opcode"
+)
+
+// interopGasRatio is a multiplier by which a number returned from price getter
+// and Fixed8 amount of GAS differ.
+const interopGasRatio = int64(100000)
+
+// gasForFree is an amount of GAS which can be spent for free = 10 GAS.
+const gasForFree = util.Fixed8(10 * 100000000)
+
+func getPrice(v *vm.VM, op opcode.Opcode, parameter []byte) int {
+	if op <= opcode.NOP {
+		return 0
+	}
+
+	switch op {
+	case opcode.APPCALL, opcode.TAILCALL:
+		return 10
+	case opcode.SYSCALL:
+		interopID := vm.GetInteropID(parameter)
+		return getSyscallPrice(v, interopID)
+	case opcode.SHA1, opcode.SHA256:
+		return 10
+	case opcode.HASH160, opcode.HASH256:
+		return 20
+	case opcode.CHECKSIG, opcode.VERIFY:
+		return 100
+	case opcode.CHECKMULTISIG:
+		estack := v.Estack()
+		if estack.Len() == 0 {
+			return 1
+		}
+
+		var cost int
+
+		item := estack.Peek(0)
+		switch item.Item().(type) {
+		case *vm.ArrayItem, *vm.StructItem:
+			cost = len(item.Array())
+		default:
+			cost = int(item.BigInt().Int64())
+		}
+
+		if cost < 1 {
+			return 1
+		}
+
+		return cost * 100
+	default:
+		return 1
+	}
+}
+
+func getSyscallPrice(v *vm.VM, id uint32) int {
+	ifunc := v.GetInteropByID(id)
+	if ifunc != nil && ifunc.Price > 0 {
+		return ifunc.Price
+	}
+
+	const (
+		neoAssetCreate           = 0x1fc6c583 // Neo.Asset.Create
+		antSharesAssetCreate     = 0x99025068 // AntShares.Asset.Create
+		neoAssetRenew            = 0x71908478 // Neo.Asset.Renew
+		antSharesAssetRenew      = 0xaf22447b // AntShares.Asset.Renew
+		neoContractCreate        = 0x6ea56cf6 // Neo.Contract.Create
+		neoContractMigrate       = 0x90621b47 // Neo.Contract.Migrate
+		antSharesContractCreate  = 0x2a28d29b // AntShares.Contract.Create
+		antSharesContractMigrate = 0xa934c8bb // AntShares.Contract.Migrate
+		systemStoragePut         = 0x84183fe6 // System.Storage.Put
+		systemStoragePutEx       = 0x3a9be173 // System.Storage.PutEx
+		neoStoragePut            = 0xf541a152 // Neo.Storage.Put
+		antSharesStoragePut      = 0x5f300a9e // AntShares.Storage.Put
+	)
+
+	estack := v.Estack()
+
+	switch id {
+	case neoAssetCreate, antSharesAssetCreate:
+		return int(5000 * 100000000 / interopGasRatio)
```

Comment:
nit: vote to have constants for all "magic numbers" in order to have documented numbers